src/watchdog.rs

```diff
@@ -0,0 +1,82 @@
+//! System Window Watchdog implementation
+
+use crate::rcc::Ccdr;
+use crate::time::Hertz;
+use cast::u8;
+use crate::hal::watchdog::{Watchdog, WatchdogEnable};
+use stm32h7::stm32h7x3::WWDG;
+
+/// Implements the System Window Watchdog
+pub struct SystemWindowWatchdog {
+    wwdg: WWDG,
+    down_counter: u8,
+    pclk3_frequency: Hertz,
+}
+
+impl SystemWindowWatchdog {
+    /// Returns a System Window Watchdog object with down_counter intialized to zero
+    /// to indicate the clock has not been used yet
+    pub fn new(wwdg: WWDG, ccdr: &Ccdr) -> Self {
+        // enable the peripheral inside the APB3
+        ccdr.rb.apb3enr.modify(|_, w| w.wwdg1en().set_bit());
+        SystemWindowWatchdog {
+            wwdg,
+            down_counter: 0,
+            pclk3_frequency: ccdr.clocks.pclk3(),
+        }
+    }
+}
+
+impl Watchdog for SystemWindowWatchdog {
+    /// Feeds the watchdog in order to avoid a reset, only executes properly if the watchdog
+    /// has already been started aka. the down_counter is not 0 anymore
+    fn feed(&mut self) {
+        // if this value is 0 it is assumed that the watchdog has not yet been started
+        assert!(self.down_counter != 0);
+        self.wwdg.cr.modify(|_, w| w.t().bits(self.down_counter));
+    }
+}
+
+impl WatchdogEnable for SystemWindowWatchdog {
+    type Time = u32;
+    /// Starts the watchdog with a given timeout period, if this period is out of bounds the function
+    /// is going to panic
+    fn start<T>(&mut self, period: T)
+    where
+        T: Into<Self::Time>,
+    {
+        let period = period.into();
+        let maximum =
+            (4096 * 2u32.pow(7) * 64) / (self.pclk3_frequency.0 / 1000);
+        assert!(period <= maximum);
+
+        // cant approximate this at compile time as the apb clock frequency is not known at compile time
+        // TODO: find a better way for this
+        let mut best_config: (u32, u32) = (0, 0);
+        let mut closest: u32 = 33334;
+        for wdgtb in 0..8 {
```

Comment:
This should be improved. Assume a constraint on T to the range [32,64), then use a match statement to find the WDGTB divisor value.

Reply:
Sorry but I'm not actually getting what you want me to do here, if I constraint T to the range of 32,64 I still gotta approximate two values for the WDGTB dvisior, one for 32 and one for 64 or not?

Reply:
To use the rust notation, constrain T[5:0] to the range `32..64` T[5:0] can't be 64. 

Something like this should do the trick. It's a common pattern in `rcc.rs`, and uses the round towards zero behaviour of integer division in rust (LLVM's `sdiv`). 
```
let ratio = period_ms * (self.pclk3_frequency.0 / 1000) / 4096;

// Prescaler
let (tb_div, tb_bits) = match ratio / 64 {
    0 => (1, 0),
    1 => (2, 1),
    2..=3 => (4, 2),
    4..=7 => (8, 3),
    // More branches here
    _ => (128, 7),
};

// Counter load value
let t = ratio / tb_div;
assert!(t < 64);

self.down_counter = u8(t).unwrap() | (1 << 6);
```

For extra clarity, use the field variants from `stm32-rs`! eg. [here](https://github.com/stm32-rs/stm32h7xx-hal/blob/master/src/rcc.rs#L516).